File: Backend/mapping/src/main/java/com/mhp/coding/challenges/mapping/mappers/ArticleMapper.java
package com.mhp.coding.challenges.mapping.mappers;

import com.mhp.coding.challenges.mapping.models.db.Article;
import com.mhp.coding.challenges.mapping.models.db.blocks.ArticleBlock;
import com.mhp.coding.challenges.mapping.models.dto.ArticleDto;
import com.mhp.coding.challenges.mapping.models.dto.blocks.ArticleBlockDto;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.ArrayList;

@Component
public class ArticleMapper {

    public ArticleDto map(Article article){
        //TODO -> done
        ArticleDto articleDto = new ArticleDto();
        
        if (article != null) {
            // use setters to map fields
            articleDto.setId(article.getId());
            articleDto.setTitle(article.getTitle());
            articleDto.setDescription(article.getDescription());
            articleDto.setAuthor(article.getAuthor());
            
            // iterate over ArticleBlocks and map each one
            ArrayList<ArticleBlockDto> blocksDto = new ArrayList<>();
            ArticleBlockDto blockDto;
            for (ArticleBlock block : article.getBlocks()) {
                blocksDto.add(block.map());
            }
            // sort the mapped ArticleBlockDto
            Collections.sort(blocksDto);
            articleDto.setBlocks(blocksDto);
        }
        
        return articleDto;
    }

    public Article map(ArticleDto articleDto) {
        // Nicht Teil dieser Challenge.
        return new Article();
    }
}

File: Backend/mapping-kotlin/src/main/kotlin/com/mhp/coding/challenges/mapping/mappers/ArticleMapper.kt
package com.mhp.coding.challenges.mapping.mappers

import com.mhp.coding.challenges.mapping.models.db.Article
import com.mhp.coding.challenges.mapping.models.dto.ArticleDto
import com.mhp.coding.challenges.mapping.models.dto.blocks.ArticleBlockDto;
import org.springframework.stereotype.Component
import java.util.*

@Component
class ArticleMapper {
    // removed "?" so article cannot be null (null-safety)
    fun map(article: Article): ArticleDto {
        //TODO -> done
        val _description : String? = article.description
        val _author : String? = article.author
        // map fields
        var articleDto = ArticleDto(
            id = article.id,
            title = article.title,
            description = if (_description != null) _description else "",
            author = if (_author != null) _author else "",
            blocks = ArrayList<ArticleBlockDto>()
        )
        
        // iterate over ArticleBlocks and map each one
        var blocksDto = ArrayList<ArticleBlockDto>()
        for (block in article.blocks) {
            blocksDto.add(block.map())
        }
        // sort the mapped ArticleBlockDto
        Collections.sort(blocksDto);
        articleDto.blocks = blocksDto
        
        return articleDto;
    }

    // Not part of the challenge / Nicht Teil dieser Challenge.
    fun map(articleDto: ArticleDto?): Article = Article(
        title = "An Article",
        blocks = emptySet(),
        id = 1,
        lastModified = Date()
    )
}
